modules/library/src/main/scala/zio/elasticsearch/ElasticResult.scala

```diff
@@ -16,34 +16,138 @@
     ZIO
       .fromEither(doc match {
         case Some(item) =>
           item.documentAs match {
-            case Left(e)    => Left(DecodingException(s"Could not parse the document: ${e.message}"))
             case Right(doc) => Right(Some(doc))
+            case Left(e)    => Left(DecodingException(s"Could not parse the document: ${e.message}"))
           }
         case None =>
           Right(None)
       })
       .mapError(e => DecodingException(s"Could not parse the document: ${e.message}"))
 }
 
-final class SearchResult private[elasticsearch] (private val hits: List[Item]) extends ElasticResult[List] {
+final class SearchResult private[elasticsearch] (private val hits: List[Item]) extends DocumentResult[List] {
   def documentAs[A: Schema]: IO[DecodingException, List[A]] =
     ZIO.fromEither {
       ZValidation.validateAll(hits.map(item => ZValidation.fromEither(item.documentAs))).toEitherWith { errors =>
         DecodingException(s"Could not parse all documents successfully: ${errors.map(_.message).mkString(",")})")
       }
     }
 }
+
+final class SearchWithAggregationsResult private[elasticsearch] (
+  private val hits: List[Item],
+  private val aggregationsJson: Option[Json]
+) extends DocumentsWithAggregationsResult {
+  def aggregation(name: String): Task[Option[ElasticAggregationResponse]] = ZIO.fromEither {
+    decodeAggregationsJson(aggregationsJson, Some(name)).map(res => res.get(name))
+  }
+
+  def aggregations: Task[Map[String, ElasticAggregationResponse]] = ZIO.fromEither {
+    decodeAggregationsJson(aggregationsJson, None)
+  }
+
+  def documentAs[A: Schema]: Task[List[A]] = ZIO.fromEither {
+    ZValidation.validateAll(hits.map(item => ZValidation.fromEither(item.documentAs))).toEitherWith { errors =>
+      DecodingException(
+        s"Could not parse all documents successfully: ${errors.map(_.message).mkString(",")})"
+      )
+    }
+  }
+}
+
+private object ElasticResult {
+  def decodeAggregationsJson(
+    aggregationsJson: Option[Json],
+    name: Option[String]
+  ): Either[DecodingException, Map[String, TermsAggregationResponse]] = {
+
+    val decodingExceptionMessage = name match {
+      case Some(_) => "Could not parse aggregation successfully."
+      case None    => "Could not parse all aggregations successfully."
+    }
+
+    def getUsefulPairs(pairs: List[(String, Json)]): List[(String, Json)] =
+      name match {
+        case Some(str) => pairs.filter { case (field, _) => field == str }
+        case None      => pairs
+      }
+
+    aggregationsJson match {
+      case Some(aggregations) =>
+        Obj.decoder.decodeJson(aggregations.toString) match {
+          case Right(res) =>
+            Validation
+              .validateAll(
+                getUsefulPairs(res.fields.toList).map { case (field, data) =>
+                  ZValidation.fromEither(
+                    field match {
+                      case str if str.contains("terms") =>
```

Comment:
Similar decoding was implemented under `ElasticAggregationResponse`, why do we need both?